MongoDB Change Streams 🌿

italojs opened the pull request:
This is the first implementation of Change Streams in Meteor’s data layer. While the core functionality is working, we welcome feedback on API design and DX improvements before finalizing.

# MongoDB Change Streams

This pull request introduces support for MongoDB Change Streams in Meteor's core, replacing the current oplog option.

Change streams are more efficient than the traditional oplog tailing approach for several reasons:

With oplog tailing, Meteor must continuously poll and process the entire MongoDB oplog, filtering out irrelevant changes for each collection and user. This results in unnecessary CPU and memory usage, especially as the number of collections, users, or writes increases.

Change streams are natively supported by MongoDB and are designed for scalability, providing a more direct and efficient notification mechanism for data changes.

We will maintain the old Oplog and Pooling implementation, so if you aren't able to use change stream or just don't want to do it, you will flag it, and it SHOULD be transparent for you; any change else will be required.

## How to use it

```
{
    "packages": {
        "mongo": {
            "reactivity": ["changeStreams", "pooling", "oplog"] // Will try using CS first, then pooling, and if that does not work, oplog
        }
    }
}
```


Note: we should move the commit [48301e6](https://github.com/meteor/meteor/pull/13787/commits/48301e60565394d34a04cb64e88e0008f4a4378d) to devel when release 3.4.1 and update this branch with the changes

radekmie commented:
I was surprised when I saw this PR - maybe someone managed to solve the discrepancy between Oplog and Change Streams! But I'm afraid it's not the case. A few notes:
* The second argument to the `changed` callback contains the new field values. But what's important, if a complex type is changed (e.g., an object), the entire object is passed there. The `updatedFields` from the Change Stream on the other hand, will present the _paths_, e.g., `{ 'emails.1.address': 'x@y.z' }`.
    * Similarly, it should contain `undefined` for all removed fields. It's easy thanks to `removedFields`. But handling all types of array modifications is _hard_, due to `truncatedArrays`.
    * It _can_ be worked around using `fullDocumentBeforeChange` (at a small performance cost). That's what [`changestream-to-redis`](https://github.com/radekmie/changestream-to-redis) is doing.
* Consider this code: `collection.find(query).observeChanges({changeStreams:true, ...})`. The difference from the usual `observeChanges` is that the former notifies you based on the _collection_ events (e.g., an insert to the collection) and the latter based on the _result set_ (e.g., a new document now matches the query).
    * A simple example is to use a `{isActive: true}` query. The former will notify you only about inserts/updates/removes of `isActive: true` documents. The latter will _also_ inform you when `isActive` of any document will change.
* I'd recommend going with async iteration (`for await..of`) rather than callbacks - it's easier to batch the results.

StorytellerCZ commented:
Just a managerial node: Since this is a feature update this should be targeted towards a feature release (3.4) instead of patch release (3.3.1).

italojs commented:
@StorytellerCZ sure thing, I opened this PR to 3.3.1 because we will release the new version this week, and we will rebase some commits from devel to release-3.3. 
If I create a new branch from release-3.3 right now, I will probably have a few conflicts later, and the same goes for devel.


I will point this PR to release 3.4 after the release 3.3 is completed. Thanks for the heads up

italojs commented:
@radekmie 
- I'm sorry, I didn't receive your badge cases. Could you provide more details or share some forum/GH discussion? 
- Do you think those scenarios apply to watchChangeStream too, since it returns a direct WS connection?

italojs commented:
Status report: 
Following [this](https://forums.meteor.com/t/mongodb-change-streams-support-in-meteor-feedback-wanted/63681/6?u=italojs) comment, I will work on observeChanges compatibility only after the new API is ready

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/meteor/meteor?pullRequest=13787) <br/>All committers have signed the CLA.

italojs commented:
@radekmie, it’s still in progress, but now I have implemented a changeLog_observe_driver, I suppose it follows your expectation about `solve the discrepancy between Oplog and Change Streams`, am I right?

The test is 100% compatible with the current oplog, and a few benchmarks I did (which I will share soon) show Meteor 15% faster. I'm still working on a scale improvement, so it's not ready for review, but I would like to get your opinion at this point.

nachocodoner commented on packages/mongo/tests/changestream_observe_driver_tests.js:
```diff
@@ -0,0 +1,123 @@
+  Tinytest.add('mongo - ChangeStreamObserveDriver - skipped (not supported)', function (test) {
+    test.isTrue(true, 'Change Streams not supported in this environment');
+  });
+}
```
It’s fine to add this new test for the change stream driver. I’m curious what would happen if we enabled change streams by default across the test suite. If the driver handles everything correctly, all tests should pass. We have many tests expecting publication setups, and the change stream must be included in a way that remains compatible with all current assertions.

To try it, I’d branch off this one and set the env var so all TravisCI test-in-console runs use this change stream driver. Maybe later we can add a CI job that runs change streams for the whole suite.

PD: Current TravisCI tests are failing.

Grubba27 commented on packages/mongo/mongo_connection.js:
```diff
@@ -850,6 +891,35 @@ Object.assign(MongoConnection.prototype, {
+        function () {
+          // Check if change streams are explicitly disabled
+          const mongoSettings = Meteor.settings?.packages?.mongo || {};
+          return mongoSettings.useChangeStreams !== false;
```
we should add this flag to the docs – I would change it to be default on for new projects, it would be nice to add it to our skeletons.

Maybe naming could be a string:


`reactivityMode` : `ChangeStream` | `Oplog`

Also, having an easy way to turn this on via env variable would be nice for testing for example, we could run the CI with both change stream and oplog to catch issues that might appear

Grubba27 commented on packages/mongo/changestream_observe_driver.js:
```diff
@@ -0,0 +1,262 @@
+            if (!self._stopped) {
+              self._restartChangeStream();
+            }
+          }, 1000);
```
Maybe move this 1000 to a variable that could be used by the dev?

Grubba27 commented on packages/mongo/changestream_observe_driver.js:
```diff
@@ -0,0 +1,262 @@
+      const selector = this._cursorDescription.selector || {};
+      const options = { ...this._cursorDescription.options };
+      
+      // Remove some options that don't apply to find()
+      delete options.tailable;
+      delete options.oplogReplay;
```
Why is it necessary to delete those keys?

Grubba27 commented on packages/mongo/changestream_observe_driver.js:
```diff
@@ -0,0 +1,262 @@
+      delete options.oplogReplay;
+      
+      // Find all existing documents
+      const cursor = collection.find(selector, options);
+      const docs = await cursor.toArray();
+      
+      // Send 'added' for each existing document that matches our matcher
+      for (const doc of docs) {
```
You can run [`for await (const doc of cursor) {`](https://www.mongodb.com/docs/drivers/node/current/crud/query/cursor/#asynchronous-iteration)


When using `toArray()` large numbers of matched documents can cause performance issues or failures if the operation exceeds memory constraints

Grubba27 commented on packages/mongo/mongo_connection.js:
```diff
@@ -33,6 +34,7 @@ export const MongoConnection = function (url, options) {
     ignoreUndefined: true,
   }, userOptions);
 
+  console.log('mongoOptions:', mongoOptions);
```
I think this should be removed

Grubba27 commented on packages/mongo/mongo_connection.js:
```diff
@@ -113,6 +117,43 @@ MongoConnection.prototype.close = function () {
+    const version = serverInfo.version.split('.').map(Number);
+    
+    // Check MongoDB version (3.6+)
+    const hasMinVersion = version[0] > 3 || (version[0] === 3 && version[1] >= 6);
```
I think we have a semver package to do this kind of operation ( would it handle versions with betas in the name?)

italojs replied:
for mongo?

radekmie replied:
I agree - manually parsing the version may be risky.

italojs replied:
The semver is inside the tool not in the app level, I think it's OK to check the version, I wouldn't like to add a new dependency to meteor app just to check it

Grubba27 commented on packages/mongo/collection/collection.js:
```diff
@@ -165,6 +164,87 @@ Object.assign(Mongo.Collection.prototype, {
     return this._connection && this._connection !== Meteor.server;
   },
 
+  /**
+   * @summary Watch MongoDB Change Streams and publish changes via DDP.
+   * @locus Server
+   * @memberof Mongo.Collection
+   * @instance
+   * @param {Object} subscription The DDP subscription object
+   * @param {Array} [pipeline] Optional aggregation pipeline to filter Change Stream events.
+   * @param {Object} [options] Optional settings for the Change Stream.
+   * @returns {Object} Handle with stop() method to stop watching.
+   * @throws {Error} If called on a client/minimongo collection.
+   *
+   * @example
+   *   // In a publish function:
+   *   Meteor.publish('myChanges', function() {
+   *     return MyCollection.watch(this, [
+   *       { $match: { 'operationType': 'insert' } }
+   *     ]);
+   *   });
+   */
+  watch(subscription, pipeline = [], options = {}) {
```
How would this snippet below work in this `watch` example?
```js
Meteor.publish('myChanges', function() {
       return MyCollection.find();
})
```

Also, passing the `this`, seems a bit weird; maybe the API could be `MyCollection.watch(pipeline, options)`. The `Publish` method could check if it is a `watch` method, and if it is, run the necessary operations

Also2, would it work like we do for [array of cursors?](https://arc.net/l/quote/ajibbhyo)

italojs replied:
for this I'm awaitting a new PR from @rodrigooler 
So disregard it for a while

Twisterking commented:
@italojs 
Are you planning to also support the exclude & include setups for changestreams?
See https://github.com/meteor/meteor/issues/13945 and https://github.com/meteor/meteor/pull/13009!

I am not deep in the topic of changestreams, but ignoring certain updates as soon as possible in the process for collections you just are "not interested in for pub/sub" I feel also makes sense for changestreams?! We should try our best to reduce CPU (and MEM) load as much as possible.
Nice job!! 🥳

italojs commented:
@Twisterking for sure, this changestreams observer driver should be the most compatible with oplog as possible in the first version.
but the performance tunning will come as a second step

italojs committed:
Fix observeChanges tests to handle change streams correctly and improve oplog tests

- Updated observeChanges tests to ensure no update events are emitted when only hidden fields are modified, specifically when using change streams.
- Refactored oplog tests to conditionally run based on the environment variable METEOR_REACTIVITY.
- Enhanced oplog tests to ensure proper handling of inclusion and exclusion of collections, preventing conflicts.
- Added assertions to verify expected behavior when documents are inserted into included and excluded collections.

italojs commented:
@radekmie, finally, we have a first change stream version. 
Would you like to review it?
The next step will be the performance improvement, but at least all tests is passing now :)

Grubba27 commented:
on the Settings maybe using just strings could be better?!

instead of 
```json
{
    "packages": {
        "mongo": {
            "reactivity": {
                "changeStreams": true,
                "oplog": false,
                "pooling": false
            }
        }
    }
}
```
this:
```json
{
    "packages": {
        "mongo": {
            "reactivity": "changeStreams" // only changeStreams
        }
    }
}
```

Or:

```json
{
    "packages": {
        "mongo": {
            "reactivity": ["changeStreams", "pooling", "oplog"] // Will try using CS first, then pooling, and if that does not work, oplog
        }
    }
}
```

radekmie commented on packages/ddp-rate-limiter/ddp-rate-limiter-tests.js:
```diff
@@ -344,20 +344,23 @@ function createTestUser(test, expect) {
+  }
 
-  return { username, email, password };
+  return ret;
```
@italojs Do you think it would be possible to merge the unrelated changes first, so this PR gets smaller? I'm having a hard time reviewing it, as it's often unclear _why_ the changes are here.

italojs replied:
👍🏻 i'll take a look on it, I remember I was having many fail tests in non-related pkgs but many was flaky tests, I extract a few changes as you can see [here](https://github.com/meteor/meteor/pull/13972) but maybe i need cut-off a litle bit more

radekmie commented on packages/minimongo/local_collection.js:
```diff
@@ -102,8 +102,46 @@ export default class LocalCollection {
+      this._serializedBinaryData(result);
+    }
+    
+    return result;
```
1. Is it correct, that it's only applied for `findOneAsync`? What about the non-async version, or the more general `.find().fetch()`? I'm asking, since:
    * It's surprising to see it in this PR.
    * It's _really_ costly, especially for "wide" documents (i.e., ones with a lot of keys).
2. There's no need to check for `object`, as `fetch()` will return an array of objects anyway.

italojs replied:
You are right; it's not necessarily required. The context here is that at the beginning of the implementation, to be honest a few months ago I was a little lost in the codebase since I just arrived at Meteor and got this challenge to solve . I added it because the tests are flaky on my machine for some reason I don't know, and this made the binary tests pass. I ran the tests here filtering by the "binary" string, and all of them passed.

radekmie commented on packages/minimongo/local_collection.js:
```diff
@@ -102,8 +102,46 @@ export default class LocalCollection {
+      if (value && typeof value === 'object' && 
+          value.sub_type !== undefined && 
+          value.buffer !== undefined && 
+          value.position !== undefined) {
```
Would `instanceof` be applicable here?

italojs replied:
removed with _serializedBinaryData

radekmie commented on packages/minimongo/local_collection.js:
```diff
@@ -1617,9 +1655,9 @@ LocalCollection._observeFromObserveChanges = (cursor, observeCallbacks) => {
-        const doc = transform(this.docs.get(id));
+        // Use docFromDriver if available, else fallback to cache
+        const cachedOrdered = this.docs.get(id) || { _id: id };
+        const doc = transform(cachedOrdered);
```
This is a _really_ old piece of code. Can you explain what triggered the issue here?

radekmie commented on packages/minimongo/local_collection.js:
```diff
@@ -1719,8 +1758,7 @@ LocalCollection._removeFromResultsSync = (query, doc) => {
     const id = doc._id;  // in case callback mutates doc
-
-    query.removed(doc._id);
+    query.removed(String(id));
```
This breaks queries of documents with non-string IDs, right?

radekmie commented on packages/minimongo/local_collection.js:
```diff
@@ -1734,7 +1772,7 @@ LocalCollection._removeFromResultsAsync = async (query, doc) => {
     const id = doc._id;  // in case callback mutates doc
 
-    await query.removed(doc._id);
+  await query.removed(String(id));
```
(minor) It'd be great to keep the formatting consistent.

radekmie commented on packages/mongo-id/id.js:
```diff
@@ -75,6 +75,12 @@ MongoID.idStringify = (id) => {
   } else if (typeof id === 'object' && id !== null) {
+    if (typeof id.toHexString === 'function') {
+      return id.toHexString();
+    }
```
This _looks_ like `ObjectID`'s method. Why isn't it handled with the `instanceof` above?

italojs replied:
i'm not using it since the package itself is a simulation of objectID and i'm not confident `_looksLikeObjectID === true` would have the HexToString function 

btw, why do we have a simulation of ObjectID instead mongo.ObjectID itself?

radekmie replied:
Then it should be extracted into a helper. I doubt there's _a lot_ of such objects, but calling such a method on _any_ object sounds risky.

> btw, why do we have a simulation of ObjectID instead mongo.ObjectID itself?

Cause the `mongodb` package cannot be used in the browser.

radekmie commented on packages/mongo-id/id.js:
```diff
@@ -75,6 +75,12 @@ MongoID.idStringify = (id) => {
+    }
+    if (typeof id.toString === 'function') {
+      return id.toString();
+    }
```
That's a risky breaking change, as custom classes often do have _some_ `toString`. They were rejected before (correctly), but will trigger some different behavior now (potentially incorrect, if their ID is not entirely relevant, e.g., summarizes the object).

radekmie commented on packages/mongo/tests/observe_changes_tests.js:
```diff
@@ -519,8 +519,8 @@ if (Meteor.isServer) {
 
-      await self.insert({x: 2, y: 3});
       self.expects.push(resolver1, resolver2);
+      await self.insert({x: 2, y: 3});
```
Does the order matter here? I think it does not, but I'm interested _why_ this change was made.

italojs replied:
The order matter because if you call `insert` before add the resolver inside the  `self.expects` array, the `added` callback breaks since `self.expects` is empty` 
<img width="619" height="165" alt="Screenshot 2025-11-10 at 11 36 10 AM" src="https://github.com/user-attachments/assets/f6a07977-5ff7-40a5-a0a1-3df4065f6134" />

maybe it doent matter for oplog due a latency issue

radekmie replied:
Then I'd extract this change to the pre-merged tests, too. Just to make sure it passed before as well.

italojs replied:
this fix comes with change streams context, since it doesnt matter for oplog it should be included in this PR IMO, to mantain the context, would be useful to know this change comes after change streams

radekmie replied:
It depends on how you look at it. I'd rather not see any test _changes_ in this PR, as this suggests some things have changed for the seemingly unchanged logic.

italojs replied:
https://github.com/meteor/meteor/pull/14018

radekmie commented on packages/mongo/mongo_common.js:
```diff
@@ -133,6 +133,18 @@ export const replaceMongoAtomWithMeteor = function (document) {
+    if (document.buffer && (document.buffer instanceof Uint8Array || Array.isArray(document.buffer))) {
+      return new Uint8Array(document.buffer);
+    }
+  }
```
This seems like a more suitable place for this conversion, but:
1. Why wasn't it necessary before?
2. Shouldn't we check for `sub_type !== 0`, just like the test above?
3. It'd be more in-align with the rest of the codebase to merge these conditions into one.

radekmie commented on packages/mongo/mongo_connection.js:
```diff
@@ -90,6 +90,8 @@ export const MongoConnection = function (url, options) {
   }
 
+  // Check if Change Streams are supported
+  self._checkChangeStreamSupport();
```
1. This is an asynchronous operation triggered in the constructor. It may result in a race condition, where `this._supportsChangeStreams` will be `false` not because it's not supported, but because the check is still ongoing.
2. This method executes normal MongoDB operations, meaning it'll force a connection. It wasn't the case before, i.e., it was _lazily_ opened, when needed.

Ideally, this would be done in `_observeChanges` and memoized for the next executions.

radekmie commented on packages/mongo/mongo_connection.js:
```diff
@@ -114,6 +116,37 @@ MongoConnection.prototype.close = function () {
+    }
+    
+    // Check if we're running on a replica set or sharded cluster
+    const isMaster = await admin.command({ isMaster: 1 });
```
Most applications are, so it'd be ideal to start this query in parallel to `serverInfo` and `await` it here, so the latency is lower.

radekmie commented on packages/mongo/mongo_connection.js:
```diff
@@ -114,6 +116,37 @@ MongoConnection.prototype.close = function () {
+    
+    // Check if we're running on a replica set or sharded cluster
+    const isMaster = await admin.command({ isMaster: 1 });
+    const isReplicaSet = isMaster.setName || isMaster.ismaster || isMaster.secondary;
```
`setName` is a `string | undefined`, not `boolean` - it'd be better to keep this as `boolean` though, not to leak such information.

radekmie commented on packages/mongo/mongo_connection.js:
```diff
@@ -114,6 +116,37 @@ MongoConnection.prototype.close = function () {
+    self._supportsChangeStreams = isReplicaSet || isSharded;
+    
+  } catch (error) {
+    console.warn('Error checking Change Streams support:', error.message);
```
I think it's more common to use `Meteor._debug` for it.

italojs replied:
didnt know about this one, thanks

radekmie commented on packages/mongo/mongo_connection.js:
```diff
@@ -852,6 +885,36 @@ Object.assign(MongoConnection.prototype, {
+          // Check if change streams are explicitly disabled
+          const mongoSettings = Meteor.settings?.packages?.mongo || {};
+          // return mongoSettings.reactivity === 'CHANGE_STREAMS' || process.env.METEOR_REACTIVITY === 'CHANGE_STREAMS';
+          return true // dumb-forcing changeStream tests into CI
```
Don't forget about this here.

radekmie commented on packages/mongo/mongo_connection.js:
```diff
@@ -852,6 +885,36 @@ Object.assign(MongoConnection.prototype, {
+        }
+      ].every(f => f());
+      
       var canUseOplog = [
```
This should not be evaluated if change streams will be used.

italojs replied:
Do you mean the last one, right? jusr removed
but it need the matcher and the _supportsChangeStreams flag since we need to check the version and if it's a replicaset or sharded cluster

radekmie replied:
No, I mean the entire oplog pipeline. There's no need to run it, if we already know we'll use changestream.

radekmie commented on packages/mongo/mongo_connection.js:
```diff
@@ -914,7 +977,22 @@ Object.assign(MongoConnection.prototype, {
+        } catch (error) {
+          console.warn('Failed to load ChangeStreamObserveDriver, falling back to oplog/polling:', error.message);
+          driverClass = canUseOplog ? OplogObserveDriver : PollingObserveDriver;
+        }
```
Is the comment outdated? There's no dynamic import here.

radekmie commented on packages/mongo/changestream_observe_driver.js:
```diff
@@ -0,0 +1,623 @@
+      // Import Minimongo locally to avoid circular dependencies
+      const { Minimongo } = require('meteor/minimongo');
+      this._matcher = new Minimongo.Matcher(this._cursorDescription.selector);
+    }
```
When is this fallback needed? Other drivers are not doing it.

radekmie commented on packages/mongo/changestream_observe_driver.js:
```diff
@@ -0,0 +1,623 @@
+    const docs = this._multiplexer?._cache?.docs;
+    const hasDoc = typeof docs?.has === 'function'
+      ? docs.has(id)
+      : typeof docs?.get === 'function' && docs.get(id) !== undefined;
```
Typing here would help, but we have too many `?.`s here I think (e.g., the first one), and the duck typing of the methods seems unexpected - why is `has` not always present?

radekmie commented on packages/mongo/changestream_observe_driver.js:
```diff
@@ -0,0 +1,623 @@
+      if (!matchesBefore) {
+        // Document wasn't previously in the result set and now matches – emit added
+        const projectedDoc = this._projectionFn ? this._projectionFn(newDoc) : newDoc;
+        this._sendMultiplexerAdded(id, projectedDoc);
```
This operation is marked as `async`. I don't think it has to, but dangling `Promise`s are not a good thing.

radekmie commented on packages/mongo/changestream_observe_driver.js:
```diff
@@ -0,0 +1,623 @@
+    // Determine which state (before/after) matches the cursor selector
+    const matchesAfter = this._matcher
+      ? this._matcher.documentMatches(newDoc || {}).result
+      : true;
```
`_matcher` should be always present.

radekmie commented on packages/mongo/changestream_observe_driver.js:
```diff
@@ -0,0 +1,623 @@
+    if (aLow > bLow) return 1;
+    if (aLow < bLow) return -1;
+    return 0;
+  }
```
1. This should not be a method, but a helper.
2. Isn't there some sort of `compare()` method provided by the driver?

italojs replied:
1. I can move it to `mongo_utils.js` since it's a function to compare 2 MongoDB Timestamps, it's a generic function but the unique propose untiil now is to use in mongo
2. Unfortinately no, let me know if you know something, for sure I want to remove that function from there.

radekmie replied:
2. I didn't _know_, there was one, but I'd be surprised if there wouldn't be any. I did a quick check and there's both [`comp`/`compare`](https://mongodb.github.io/node-mongodb-native/6.21/classes/BSON.Timestamp.html#compare) as well as a few shorthands like [`lessThan`](https://mongodb.github.io/node-mongodb-native/6.21/classes/BSON.Timestamp.html#lessThan).

italojs replied:
good to see a doc about it but I tried it before, so since it was comming undefined i just give up, will investigate why `compare` is undefined
<img width="390" height="82" alt="Screenshot 2025-11-11 at 3 40 00 PM" src="https://github.com/user-attachments/assets/3f653090-f24d-461e-89a7-c6c60554241f" />

radekmie replied:
That's a method, not class field.

radekmie commented on packages/mongo/changestream_observe_driver.js:
```diff
@@ -0,0 +1,623 @@
+      return; // Ignore unsupported operations
+    }
+
+    const id = typeof documentKey._id !== 'string' ? new MongoID.ObjectID(documentKey._id.toHexString()) : documentKey._id;
```
`_id` can be of any arbitrary type - it has to be checked first, before calling `toHexString` on it.

radekmie commented on packages/mongo/changestream_observe_driver.js:
```diff
@@ -0,0 +1,623 @@
+      // Close current stream using stop callbacks if they exist
+      if (this._changeStream) {
+        // Find and execute the change stream stop callback
+        const changeStreamCallback = this._stopCallbacks.find(cb => 
+          cb.toString().includes('_changeStream')
+        );
```
It's stringifying functions to see if they refer to a field...? That's bold and IMO way too hacky.

italojs replied:
you are right, I admit, that is a unreviewed code from copilot

radekmie commented on packages/mongo/changestream_observe_driver.js:
```diff
@@ -0,0 +1,623 @@
+    if (aLow < bLow) return -1;
+    return 0;
+  }
+
+  _setLastProcessedOperationTime(ts) {
+    this._lastProcessedOperationTime = ts;
+    // Resolve any waiters whose target is <= current processed time
+    while (this._catchingUpResolvers.length > 0) {
+      const first = this._catchingUpResolvers[0];
+      if (this._compareOperationTimes(ts, first.ts) >= 0) {
+        this._catchingUpResolvers.shift();
+        try { first.resolver(); } catch (e) { /* ignore resolver errors */ }
+      } else {
+        break;
+      }
+    }
+  }
+
+  async _getServerOperationTime() {
+    try {
+      // Prefer a cheap ping which usually returns $clusterTime
+      const res = await this._mongoHandle.db.command({ ping: 1 });
+      return res?.operationTime || res?.$clusterTime?.clusterTime || null;
+    } catch (e) {
+      // Fallback to admin hello/ismaster
+      try {
+        const admin = this._mongoHandle.db.admin();
+        const hello = await admin.command({ hello: 1 });
+        return hello?.operationTime || hello?.$clusterTime?.clusterTime || null;
+      } catch (e2) {
+        try {
+          const admin = this._mongoHandle.db.admin();
+          const im = await admin.command({ ismaster: 1 });
+          return im?.operationTime || im?.$clusterTime?.clusterTime || null;
+        } catch (e3) {
+          return null;
+        }
+      }
+    }
+  }
```
The error handling should differentiate invalid/unsupported commands and other errors (e.g., connection failure).

radekmie commented on packages/mongo/changestream_observe_driver.js:
```diff
@@ -0,0 +1,623 @@
+        collMod: collectionName,
+        changeStreamPreAndPostImages: { enabled: true }
+      });
+    }
```
This is opinionated, but I'd find it surprising that an _observe_ driver _creates_ a collection. Especially, since it will assume some settings, e.g., no collation.

italojs replied:
I did it since the collections must to have `changeStreamPreAndPostImages`  enabled to give me the `fullDocumentBeforeChange`, where I use to handle the the diff for update since Meteor Multiplexer just require the changes.

But looking again on it, I'll see if it's possible to use the changes that comes from change streams only 🤔

radekmie replied:
You can always start a change stream and only report that this option is not enabled _while processing the events_. Sure, it'll get reported later, but it brings less friction. (It's also safer, as each of the commented operations can fail.)

radekmie commented on packages/mongo/changestream_observe_driver.js:
```diff
@@ -0,0 +1,623 @@
+
+      // Create change stream with appropriate options
+      const changeStreamOptions = {
+        fullDocument: 'updateLookup',
+        fullDocumentBeforeChange: 'whenAvailable'
```
These two should be configurable, as `updateLookup` has _some_ delay, and `required` is not always possible.

italojs replied:
I was planning to address it as a patch 3.5.1 (pr at least use it for the beta), I would like to restrict multiple config scenarios in a beta or in the first version

radekmie reviewed (commented):
Here's my initial, **really** not in-detail review. Overall, there's a lot of uncovered edge cases (e.g., `console.warn`s or silent `catch` blocks), unexpected changes (see the first comment), and a lot of tests I haven't reviewed yet at all.

And one more question: the change stream is always opened with a very basic `$match` - that's clear. But I couldn't find any multiplexing of the observers themselves, and that'd mean every one of them would create a new change stream, which is far from ideal.

italojs committed:
Remove recursive serialized-binary conversion from findOneAsync

Return the fetched document directly in findOneAsync and drop the
_serializedBinaryData helper. This removes the previous recursive attempt
to convert serialized MongoDB.Binary-like objects into Uint8Array, avoids
mutating returned documents, and simplifies the async findOne code path.

italojs committed:
Only run oplog tests when not using change streams, skip flaky nested test loop

Negate testingChangeStream checks so oplog/oplog-related tests execute only
in non-change-stream mode. Add an early return to disable a flaky nested
repetition loop in the test matrix. Add a TODO comment in runInFence about
removing the temporary timeout.

italojs committed:
Add compareOperationTimes helper and switch ChangeStreamObserveDriver to use it; standardize oplog test gating

- Add compareOperationTimes to packages/mongo/mongo_utils.js for comparing MongoDB clusterTime/timestamp shapes.
- Replace internal _compareOperationTimes usages in changestream_observe_driver.js with the new shared helper.
- Update packages/mongo/tests/mongo_livedata_tests.js to consolidate test gating under testingOplog, remove the change-stream-specific delay, and flip conditionals to only run oplog tests when testingOplog is enabled.

italojs committed:
Use Meteor._debug instead of console.warn for MongoConnection warnings and ChangeStream load failure; remove redundant ChangeStream cursorSupported check from observe gating; temporarily disable flaky native-find maxTimeMs test; fix indentation in LocalCollection._removeFromResultsAsync

italojs committed:
fixing versioning

italojs committed:
ChangeStreamObserveDriver: remove async from _sendMultiplexerAdded; improve stop-callback detection; normalize _id handling; treat replace as update

- Make _sendMultiplexerAdded synchronous (remove unnecessary async).
- Replace fragile cb.toString() heuristic with a property check (typeof cb._changeStream === 'function') when locating the change stream stop callback.
- Construct id from documentKey._id more robustly by converting ObjectId-like values via toHexString().
- Ensure 'replace' is handled alongside 'update' in the pending-writes switch so replace events are treated like updates.

italojs committed:
ChangeStreamObserveDriver: treat replace as update; remove redundant _handleReplace and duplicate switch branch

italojs committed:
ChangeStreamObserveDriver: remove internal matcher fallback and _compareOperationTimes; mongo-id: use _looksLikeObjectID in idStringify; mongo: log observe driver selection

italojs committed:
Merge branch 'devel' into feat/change-streams

italojs committed:
ChangeStreams: enforce unordered observes; improve opTime retrieval, error handling and logging; re-enable tests

- ChangeStreamObserveDriver: throw if options.ordered (only unordered observeChanges supported)
- Wrap multiplexer.added in try/catch and log errors to avoid crashing on client delivery failures
- Refactor _getServerOperationTime to try db.command({ping}), admin.command({hello}), admin.command({ismaster}) in sequence; handle unsupported command errors and log failures
- Remove internal matcher fallbacks in _handleInsert/_handleUpdate (assume matcher is present) and simplify _handleDelete cache check
- MongoConnection: parse server version robustly (major/minor), require MongoDB 3.6+ using major/minor comparison, coerce isReplicaSet to Boolean, and gate change-stream selection on _supportsChangeStreams
- Add logging when selector compilation for change streams fails
- Tests: re-enable previously commented collection test and remove accidental early return that skipped livedata test permutations

italojs committed:
ChangeStreams: inline opTime comparison into ChangeStreamObserveDriver; remove duplicate compareOperationTimes from mongo_utils; reuse isMaster promise and drop debug log in MongoConnection; LocalCollection: avoid passing placeholder {_id} to transform in _observeFromObserveChanges

italojs committed:
mongo: remove fallback for serialized MongoDB.Binary plain-object conversion

Remove the special-case detection/conversion of plain objects with sub_type/buffer/position
back to Uint8Array in replaceMongoAtomWithMeteor. This fallback was unnecessary and could
cause incorrect conversions.

italojs commented:
forgot to mention, i'm off for a few days off, will be back in 2 weeks

italojs committed:
mongo: refactor observe driver selection and reactivity configuration

- Replace ad-hoc driver checks with a configurable selection pipeline:
  - support Meteor.settings.packages.mongo.reactivity as a string or array
  - map driver names to implementations and validate configured names
  - allow custom driver order (fall back to default ['oplog','changeStreams','pooling'])
- Implement per-driver availability checks that return structured reasons and matcher/sorter objects
  - clearer handling for changeStreams, oplog, and polling/pooling fallbacks
  - surface helpful error messages when no driver can be selected
- Defer Change Streams capability detection (lazy _checkChangeStreamSupport) and fix this/self usage in that helper
- Remove eager _checkChangeStreamSupport call from constructor
- Emit a log with the selected driver and collection for easier debugging
- Normalize 'pooling' → PollingObserveDriver and improve oplog include/exclude warnings

italojs committed:
mongo: move compareOperationTimes into mongo_common; ChangeStreams: import and use it and remove local fallback

italojs committed:
mongo: register expects after first insert in tailable observe test to avoid race

Move self.expects.push(resolver1, resolver2) to after the initial insert in
the tailable observeChanges test so the inserted event doesn't fire before
the test has registered its expectations.

italojs committed:
mongo: allow forcing reactivity via METEOR_REACTIVITY; simplify driver order logic; restore debug log; comment out ordered-only ChangeStreams restriction

italojs committed:
mongo: make reactivity order configurable and prefer Change Streams; remove METEOR_REACTIVITY override; restore debug log; re-enable ordered-only ChangeStreams restriction; tidy observe driver args and test guards

italojs committed:
Merge branch 'devel' into feat/change-streams

italojs committed:
ddp-rate-limiter: simplify createTestUser; call Accounts.createUser directly and remove conditional logout/doCreate

italojs committed:
Ci/removing travis (#14026)

* Add travis-compat GitHub Actions workflow to run Travis parity tests

- Add .github/workflows/travis-compat.yml
- Trigger on push to ci/removing-travis
- Runs on ubuntu-22.04 with Node 22.17 and C++ toolchain g++-12
- Restores caches (.npm, .meteor, .babel-cache, dev_bundle, chromium), installs system/npm deps and runs packages/test-in-console/run.sh

* Install Puppeteer in test-packages workflow, add jquery to test-in-console, and export METEOR_NO_DEPRECATION earlier in run.sh

* Detect writable group dynamically in webapp socket tests and skip when none found

Add getGroupNameForGid and getWritableGroupName helpers to resolve a writable
group from /etc/group, the current user's gid/groups, or fallbacks (TRAVIS/staff/root).
Use the resolved group for UNIX_SOCKET_GROUP in socket_file_tests and skip the
group-specific test when no writable group can be determined.

* Update app.md

Fix example value for a preference

* Rename workflow job, remove Puppeteer install, add TINYTEST_FILTER, and add email debug logs

- Rename GitHub Actions job travis-compat -> test-packages
- Add TINYTEST_FILTER="email" to test-packages workflow env
- Remove global Puppeteer install step and drop "Travis parity" label from run step
- Add console.log debug output in Email.sendAsync to show MAIL_URL env, settings, and Meteor.isProduction before the production check

* Remove console.log debug output and include MAIL_URL, settings, and production flag in missing-mail error message

* Unquote TINYTEST_FILTER value and add NODE_ENV: CI to test-packages workflow

* Remove MAIL_URL, settings and production flag from missing-mail error message

Previously the thrown error included the MAIL_URL environment value, package
settings JSON, and Meteor.isProduction flag. That could leak sensitive
configuration into logs. Replace it with the original concise guidance and
link to the docs.

* Remove TINYTEST_FILTER from test-packages workflow

* Disable ambiguous error messages in password tests (#14027)

Set Accounts._options.ambiguousErrorMessages = false in password_tests.js
so tests receive specific, deterministic error reasons instead of ambiguous
messages.

---------

Co-authored-by: imongithubnow <76787178+imongithubnow@users.noreply.github.com>
Co-authored-by: Nacho Codoñer <igcogi@gmail.com>

italojs committed:
ci: run test-packages workflow on pull_request only; remove push branch filter and workflow name

italojs committed:
chore: add METEOR_REACTIVITY_ORDER to test workflow; fix removed id handling; minor docs & whitespace tidy

- Add workflow name/ENV and set METEOR_REACTIVITY_ORDER in .github/workflows/test-packages.yml
- LocalCollection: pass native id to query.removed (remove String coercion)
- Update app docs example BackgroundColor to use 0x-prefixed value
- Trim stray trailing whitespace/blank lines and remove extra space in ddp-rate-limiter test

italojs committed:
chore: use matrix METEOR_REACTIVITY_ORDER in CI; remove Travis; misc mongo/minimongo/id/test fixes

- CI: set METEOR_REACTIVITY_ORDER from the test-packages job matrix
- CI: remove legacy .travis.yml
- mongo: silence debug log when selecting reactivity driver
- mongo-id: tighten idStringify to reject non-ObjectID objects (remove fallback to toHexString/toString)
- minimongo: minor whitespace cleanup in _removeFromResultsSync
- tests: fix observe_changes race by registering expected callbacks before inserting document

italojs committed:
mongo: comment out automatic collection creation and collMod that enabled changeStreamPreAndPostImages in ChangeStreamObserveDriver

italojs committed:
Merge branch 'devel' into feat/change-streams

italojs committed:
mongo: fix cachedDoc access in ChangeStreamObserveDriver; add test delay to avoid flake

Use direct Map.get(id) to retrieve cached docs (avoid optional-chaining call)
and add a short Meteor._sleepForMs(100) in the unordered basics test to allow
the changeStream-driven multiplexer cache to settle and prevent timing races.

Grubba27 commented on packages/mongo/tests/observe_changes_tests.js:
```diff
@@ -188,6 +188,7 @@ Tinytest.addAsync('observeChanges - unordered - basics', async function(
       ]);
 
       await c.updateAsync(fooid, { noodles: 'alright', potatoes: 'tasty', apples: 'ok' });
+      await Meteor._sleepForMs(100); // changeStream is faster than oplog, so we need to wait a bit to update the internal multiplexer.cache
```
Could we flush the internal `multiplexer`? so that we would only wait the necessary amount of time?

italojs replied:
yes, i put it as a TODO before you comment but didnt push yet

italojs committed:
mongo: remove ordered enforcement from ChangeStreamObserveDriver; revert reactivity order; update tests; add travis.yml

- Remove storing and runtime check of options.ordered in ChangeStreamObserveDriver (no longer enforce ordered observeChanges).
- Revert DEFAULT_REACTIVITY_ORDER to ['oplog','changeStreams','pooling'] so oplog remains the default unless overridden by METEOR_REACTIVITY_ORDER.
- Replace legacy testingOplog flag with IS_OPLOG (derived from METEOR_REACTIVITY_ORDER) and gate oplog tests accordingly across tests (mongo_livedata_tests, oplog_tests).
- Small test comment/todo in observe_changes_tests to note changeStream vs oplog timing.
- Add travis.yml CI config file.

italojs committed:
mongo: inline and cache Change Stream support checks; remove helper

- Remove MongoConnection._checkChangeStreamSupport and inline the server capability
  checks into the changeStreams driver availability probe.
- Lazily compute and cache self._supportsChangeStreams and
  self._changeStreamServerReasons so the server check is only run once and
  readable reasons are preserved.
- Return early with a clear unavailable reason when change streams aren't
  supported or when ordered/_testOnlyPollCallback constraints apply.
- Improve handling of selector matcher errors by returning an unavailable reason
  instead of pushing to a shared reasons array.
- Update tests to use the new cached support fields instead of calling the
  removed helper.

italojs committed:
mongo: update multiplexer cache on change events; remove artificial test sleep

Update ChangeStreamObserveDriver to write the new document into the multiplexer
cache when emitting a 'changed' event so the cache remains in sync with change
stream updates. Also remove a 100ms test sleep from observe_changes_tests (was
used to wait for cache updates) since change streams now update faster.

italojs committed:
mongo: use actual cached doc for removed callback and preserve id type in removal

Pass the stored document directly to the observeChanges `removed` callback
instead of creating a synthetic `{_id: id}` fallback. Also stop coercing
the id to a string when awaiting `query.removed` in `_removeFromResultsAsync`
so the original id type is preserved.

italojs committed:
travis: add .travis.yml to run test suite on Node 22 (jammy)

italojs commented:
Hi @radekmie 
Addressed 99% of your requests, just is missing one, feel free to review it again

NOTES:  
1. I pushed a few changes that you can ignore while we haven't release 3.4.1:
- [test-packages.yml‎](https://github.com/meteor/meteor/pull/13787/files#diff-ebb7704d4a6cfe36b4c368e7e0ac1da7005b144e0cd0e6d59020c32fe0a72095)
- [password_tests.js‎](https://github.com/meteor/meteor/pull/13787/files#diff-862659c46eaabf2b23cb75de8905d3bb83cbd6b67ce2e95d69fb684f5001b7b5)
- "observeChanges - unordered - enters and exits result set through change" test at [observe_changes_tests.js](https://github.com/meteor/meteor/pull/13787/files#diff-fe7cf5b1fea674f63848e7beac75964fb3bad09676e5efd4c4718adc2f4a693e)
- [socket_file_tests.js](https://github.com/meteor/meteor/pull/13787/files#diff-f0d4007bf1dcb9b5efd465ff7111ba5863f6abe35837d538887c20d768288403)

All these changes I will remove when I merge it to devel (and rebase it here) in release 3.4.1

2. I ignored the changestreams unit tests for a while, I want to ensure we wont have significant changes like this last review, maybe after your next review, I will work on it

italojs committed:
minimongo: clarify comment about cloning cached doc in _observeFromObserveChanges; travis: remove trailing blank line in .travis.yml

italojs committed:
docs: add Change Streams Observer Driver performance page and sidebar entry

Add a new performance doc describing the Change Streams–based observer driver,
including requirements, configuration options, tuning tips, and performance
notes. Also add a "Change Streams Observer Driver" entry to the Performance
section in the VitePress sidebar.

radekmie commented on packages/mongo/mongo_common.js:
```diff
@@ -168,3 +168,7 @@ export function replaceNames(filter, thing) {
   return thing;
 }
+
+export function compareOperationTimes(opTime1, opTime2) {
+  return (new MongoDB.Timestamp(opTime1)).compare(opTime2);
+}
```
If we create a `Timestamp` instance out of the first argument, what are the expected types here? It'd be great to add at least a JSDoc for them.

radekmie commented on packages/mongo/mongo_connection.js:
```diff
@@ -850,78 +850,237 @@ Object.assign(MongoConnection.prototype, {
+        changeStreams: ChangeStreamObserveDriver,
+        oplog: OplogObserveDriver,
+        polling: PollingObserveDriver,
+        pooling: PollingObserveDriver,
```
Are we sure there's a need to support a typo in the options? 😅

Also, this piece of code grew significantly, and it wasn't short before either. How about we extract it into a `getObserveDriver` helper?

radekmie commented on packages/mongo/mongo_connection.js:
```diff
@@ -850,78 +850,237 @@ Object.assign(MongoConnection.prototype, {
+      const driverChecks = {
+        changeStreams: async () => {
+          let localMatcher;
+          const reasons = [];
```
The idea of `reasons` while nice, will result in an increased memory pressure in a performance-critical piece of code. I think it'd make more sense to accept the fact that there will be a lot of early-exits in this code. (It may be better with this code extracted from here.)

italojs replied:
correct me if i/m wrong, i realized we should check for _supportsChangeStreams once, since once we startup the app or first observe and _supportsChangeStreams === true, it will be rare for it became false
I'm thinking of calling it only in the first observeDriver call/creation (or in the build maybe?)

radekmie replied:
Once is enough. And it cannot happen during build - there's no MongoDB instance configured there.

radekmie commented on packages/mongo/mongo_connection.js:
```diff
@@ -850,78 +850,237 @@ Object.assign(MongoConnection.prototype, {
+      if (!driverClass) {
+        const errorDetails = availabilityErrors.length
+          ? ` Reasons: ${availabilityErrors.join(' | ')}`
+          : '';
+
+        throw new Error(`Unable to select a Mongo reactivity driver from configuration [${configuredOrder.join(', ')}].${errorDetails}`);
+      }
```
Previously this code would never throw, and I think it's a nice property to preserve. The checks for correctness of this configuration should happen on the app's start time.

radekmie commented on v3-docs/docs/performance/change-streams-observer-driver.md:
```diff
@@ -0,0 +1,64 @@
+
+- `delay.error`: Milliseconds to wait before restarting the stream after an error (default: `100`).
+- `delay.close`: Milliseconds to wait before restarting after an unexpected close (default: `100`).
+- `waitUntilCaughtUpTimeoutMs`: Upper bound for waiting until the stream catches up to the server's current operation time when coordinating with DDP fences (default: `1000`).
```
It should be well documented what can happen if this timeout is hit.

radekmie commented on v3-docs/docs/performance/change-streams-observer-driver.md:
```diff
@@ -0,0 +1,64 @@
+
+## Performance Notes
+
+- The driver requests `fullDocument: 'updateLookup'`, so updates trigger an extra read; keep indexes healthy to avoid slow lookups.
```
Indexes have nothing to do in here - it's always an `_id`-based lookup. But it _may_ happen, that some changes won't be visible, e.g., if two writes will happen almost at the same time (only the last one will be acknowledged).

radekmie commented on v3-docs/docs/performance/change-streams-observer-driver.md:
```diff
@@ -0,0 +1,64 @@
+## Performance Notes
+
+- The driver requests `fullDocument: 'updateLookup'`, so updates trigger an extra read; keep indexes healthy to avoid slow lookups.
+- Enabling MongoDB `changeStreamPreAndPostImages` on a collection lets the driver diff updates more efficiently; without it, some updates may send the full document.
```
Have you measured it? I see it falls back to whatever's in the multiplexer, which should be up to date anyway.

italojs replied:
I removed this feature from CS in your last comments, will remove this doc session

radekmie commented on v3-docs/docs/performance/change-streams-observer-driver.md:
```diff
@@ -0,0 +1,64 @@
+
+- The driver requests `fullDocument: 'updateLookup'`, so updates trigger an extra read; keep indexes healthy to avoid slow lookups.
+- Enabling MongoDB `changeStreamPreAndPostImages` on a collection lets the driver diff updates more efficiently; without it, some updates may send the full document.
+- The Change Stream pipeline currently filters only by operation type; the app-side matcher applies your selector. Very high-write collections with broad selectors may produce more events than oplog tailing.
```
Selectors have nothing to do here: the way it's done now, _all_ changes are processed, just like in the oplog implementation.

radekmie reviewed (commented):
For now I reviewed everything except for the tests and I'll need more time for that.

italojs committed:
mongo: validate and refactor reactivity driver selection

- Introduce availableDrivers and DEFAULT_REACTIVITY_ORDER (uses 'polling').
- Validate Meteor.settings.packages.mongo.reactivity if provided.
- Add _getConfiguredReactivityOrder to normalize/validate configured order.
- Extract _selectReactivityDriver to centralize async availability checks and selection.
- Simplify driverChecks and remove duplicated selection logic in _observeChanges.
- Use this consistently (instead of self) and fix multiplexer onStop closure.
- Ensure observe driver receives correct mongoHandle and store multiplexers on this.
- Add JSDoc for compareOperationTimes in mongo_common.js.

italojs committed:
mongo: remove commented collection pre/post-images setup in ChangeStreamObserveDriver; docs: remove Performance Notes and clarify waitUntilCaughtUp timeout behavior

italojs committed:
Merge branch 'release-3.5' into feat/change-streams

italojs committed:
mongo: fall back to polling when no reactivity driver selected; tests(changestream): refactor and expand ChangeStreamObserveDriver tests

- mongo_connection.js: ensure a polling driver fallback and log a debug message when no driverClass is returned from _selectReactivityDriver.
- tests/changestream_observe_driver_tests.js: complete rewrite and modernization of tests:
  - Replace old server/admin version checks with a simple driver-detection helper.
  - Convert tests to async, add helpers (makeCollection, waitFor, getPromiseAndResolver).
  - Remove brittle isMaster/version logic and server-only gating.
  - Add/clarify numerous tests covering detection, CRUD, projections, selectors, multiple observers, fence sync, pending writes, ordering, error recovery, stop/cleanup, matcher edge cases, and internal driver behaviors.
  - Improve assertions and stability (timeouts, waits) and simplify teardown.

italojs committed:
mongo: apply cache changes synchronously in ObserveMultiplexer; remove redundant cache write from ChangeStreamObserveDriver

Ensure ObserveMultiplexer updates its LocalCollection cache synchronously before scheduling async notifications to handlers to avoid races (e.g. an update arriving before the insert is recorded). Remove the manual cache set in ChangeStreamObserveDriver._handleUpdate since the multiplexer now guarantees synchronous cache updates.

harryadel commented:
:open_mouth: 
This is already tested and clear @italojs ?
